psi4/src/psi4/libfock/DFJCOSK.cc

```diff
@@ -0,0 +1,1014 @@
+            H[jki]->add(GT[jki][thread]);
+        }
+        C_DGESV(nbf_aux, 1, J_metric_->clone()->pointer()[0], nbf_aux, ipiv.data(), H[jki]->pointer(), nbf_aux);
+    }
+
+
+    // I believe C_DSYSV should be faster than C_GESV, but I've found the opposite to be true.
+    // This performance issue should be investigated, but is not consequential here.
+    // The cost of either linear solve is dwarfed by the actual integral computation.
+    //
+    //std::vector<double> work(3 * nbf_aux);
+    //int errcode = C_DSYSV('U', nbf_aux, 1, J_metric_->clone()->pointer()[0], nbf_aux, ipiv.data(), H[jki]->pointer(), nbf_aux, work.data(), 3 * nbf_aux);
+
+    // shell maxima of H for screening
+    Vector H_shell_max(nshell_aux);
+    auto H_shell_maxp = H_shell_max.pointer();
+
+    for(size_t jki = 0; jki < njk; jki++) {
+
+        auto Hp = H[jki]->pointer();
+
+        for (int P = 0; P < nshell_aux; P++) {
+            int np = auxiliary_->shell(P).nfunction();
+            int pstart = auxiliary_->shell(P).function_index();
+            for (int p = pstart; p < pstart + np; p++) {
+                H_shell_maxp[P] = std::max(H_shell_maxp[P], std::abs(Hp[p]));
+            }
+        }
+
+    }
+
+    timer_off("Metric");
+
+    //  => Third Contraction <= //
+
+    // contract H with three-index DF ERIs to get J
+    // J_{mn} = H_{p} (mn|p)
+
+    timer_on("ERI2");
+
+#pragma omp parallel for schedule(guided) num_threads(nthreads_) reduction(+ : computed_triplets2)
+    for (size_t MNP = 0; MNP < nshelltriplet; MNP++) {
+
+        size_t MN = MNP % nshellpair;
+        size_t P = MNP / nshellpair;
+        auto bra = eri_computers_[0]->shell_pairs()[MN];
+        size_t M = bra.first;
+        size_t N = bra.second;
+        if(H_shell_maxp[P] * H_shell_maxp[P] * J_metric_shell_diag[P] * eri_computers_[0]->shell_pair_value(M,N) < thresh2) {
+            continue;
+        }
+        computed_triplets2++;
+        int np = auxiliary_->shell(P).nfunction();
+        int pstart = auxiliary_->shell(P).function_index();
+        int nm = primary_->shell(M).nfunction();
+        int mstart = primary_->shell(M).function_index();
+        int nn = primary_->shell(N).nfunction();
+        int nstart = primary_->shell(N).function_index();
+        int rank = 0;
+#ifdef _OPENMP
+        rank = omp_get_thread_num();
+#endif
+
+        eri_computers_[rank]->compute_shell(P, 0, M, N);
+        const auto & buffer = eri_computers_[rank]->buffers()[0];
+
+        for(size_t jki = 0; jki < njk; jki++) {
+
+            auto JTp = JT[jki][rank]->pointer();
+            auto Hp = H[jki]->pointer();
+
+            for (int p = pstart, index = 0; p < pstart + np; p++) {
+                for (int m = mstart; m < mstart + nm; m++) {
+                    for (int n = nstart; n < nstart + nn; n++, index++) {
+                        JTp[m][n] += buffer[index] * Hp[p];
+                        if (N != M) JTp[n][m] += buffer[index] * Hp[p];
```

Comment:
Same question as before - is this thread-safe?

Reply:
Yes, each thread has its own J buffer, `JT`. That might not be ideal from a memory standpoint, but it makes the parallelism very simple.

Reply:
Question answered above. Consider this resolved!

Reply:
And for whatever it's worth, I've seen (and used) a lot of thread-private Fock buffers in the past, so I am not at all opposed to thread-private J buffers as used here.